src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -161,7 +182,9 @@ private void AllocateBuffer()
             //   and start reading into _current. Set _index to 0.
             // - When _current runs out of space, add it to _buffers and repeat the
             //   above step, except with _current.Length * 2.
+            // - Make sure we never pass _maxCapacity in all of the above steps.
 
+            Debug.Assert((uint)_maxCapacity > (uint)_count);
```

Comment:
Also, are you saying that there is a legitimate case at run-time where this assert could fail?  If so, that shouldn't be an assert.